WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,767 @@
+         field of the requested token */
+      ucEntryIndex = unToken.stFields.overlayGroupID;
+      do
+      {
+         /* search for the token */
+         pAddress = comrvSearchForLoadedOverlayGroup(pOverlayMultiGroupTokensTable[ucEntryIndex++]);
+      /* continue the search as long as the group wasn't found and we have additional tokens */
+      } while ((pAddress == NULL) && (pOverlayMultiGroupTokensTable[ucEntryIndex].uiValue != 0));
+
+      /* save the selected multi group entry */
+      usSelectedMultiGroupEntry = ucEntryIndex-1;
+   }
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+   /* get the group size */
+   usOverlayGroupSize = M_COMRV_GET_OVL_GROUP_SIZE(unToken);
+
+   /* if the data/function is not loaded we need to evict and load it */
+   if (pAddress == NULL)
+   {
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+      /* if the requested token is a multi-group token */
+      if (unToken.stFields.multiGroup != 0)
+      {
+         /* for now we take the first token in the list of tokens */
+         unToken = pOverlayMultiGroupTokensTable[unToken.stFields.overlayGroupID];
+         /* save the selected multi group entry */
+         usSelectedMultiGroupEntry = unToken.stFields.overlayGroupID;
+      }
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+      /* disable ints */
+      // TODO: disable ints
+
+      /* get eviction candidates according to the requested pOverlayGroupSize */
+      ucNumOfEvictionCandidates = comrvGetEvictionCandidates(usOverlayGroupSize, ucEvictCandidateList);
+
+      ucEntryIndex = 0;
+      /* we need to handle heap fragmentation since we got more
+         than 1 eviction candidate */
+      if (ucNumOfEvictionCandidates > 1)
+      {
+         /* no handling to the last eviction candidate */
+         ucNumOfEvictionCandidates--;
+         while (ucEntryIndex < ucNumOfEvictionCandidates)
+         {
+            /* get the candidate entry */
+            pEntry = &stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]];
+            /* calc the source address */
+            pAddress = ((u08_t*)pEntry->pFixedEntryAddress + M_COMRV_GET_OVL_GROUP_SIZE_IN_BYTES(pEntry->unToken));
+            /* perform code copy */
+            comrvMemcpyHook(pEntry->pFixedEntryAddress, pAddress,
+                  stComrvCB.stOverlayCache[ucEvictCandidateList[ucEntryIndex]].pFixedEntryAddress - pAddress);
+
+            /* after code copy we need to align the entries structures */
+            for(ucIndex = ucEvictCandidateList[ucEntryIndex] ; ucIndex < ucEvictCandidateList[ucEntryIndex+1] ; ucIndex++)
+            {
+               pEntry = &stComrvCB.stOverlayCache[ucIndex + stComrvCB.stOverlayCache[ucIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits];
+#ifdef D_COMRV_OVL_DATA_SUPPORT
+               /* an overlay data is present when handling defragmentation */
+               if (pEntry->unProperties.ucData)
```

Comment:
do in assert macro

Reply:
can't have here assert since in the next line we have M_COMRV_EXIT_CRITICAL_SECTION();